features/distributed/coordination/zookeeper/src/main/java/org/opennms/features/distributed/coordination/zookeeper/ZookeeperDomainManager.java

```diff
@@ -0,0 +1,223 @@
+     *
+     * @param domain        the domain to manage
+     * @param clientBuilder the curator framework client to use
+     */
+    private ZookeeperDomainManager(String domain, CuratorFrameworkFactory.Builder clientBuilder) {
+        super(domain);
+        this.clientBuilder = clientBuilder;
+    }
+
+    /**
+     * Private constructor to be used for unit testing only.
+     *
+     * @param domain             the domain
+     * @param mockClientBuilder  a mocked client
+     * @param mockLeaderSelector a mocked leader selector
+     */
+    private ZookeeperDomainManager(String domain, CuratorFrameworkFactory.Builder mockClientBuilder,
+                                   LeaderSelector mockLeaderSelector) {
+        super(domain);
+        clientBuilder = mockClientBuilder;
+        leaderSelector = mockLeaderSelector;
+    }
+
+    /**
+     * Default static factory method.
+     *
+     * @param domain        the domain to manage
+     * @param clientBuilder the client to use
+     * @return the manager instance
+     */
+    static ZookeeperDomainManager of(String domain, CuratorFrameworkFactory.Builder clientBuilder) {
+        return new ZookeeperDomainManager(domain, clientBuilder);
+    }
+
+    /**
+     * Static factory method for use in unit tests using mocks.
+     *
+     * @param domain             the domain to manage
+     * @param mockClientBuilder  the mock client
+     * @param mockLeaderSelector the mock leader selector
+     */
+    static ZookeeperDomainManager withMocks(String domain, CuratorFrameworkFactory.Builder mockClientBuilder,
+                                            LeaderSelector mockLeaderSelector) {
+        return new ZookeeperDomainManager(domain, mockClientBuilder, mockLeaderSelector);
+    }
+
+    /**
+     * Closes the connection.
+     */
+    private void closeConnection() {
+        if (leaderSelector != null) {
+            leaderSelector.close();
+            leaderSelector = null;
+        }
+
+        client.close();
+        client = null;
+    }
+
+    /**
+     * A method for unit testing to obtain the {@link LeaderSelectorListener} being used.
+     *
+     * @return the LeaderSelectorListener
+     */
+    LeaderSelectorListener getLeaderSelectorListener() {
+        return leaderSelectorListener;
+    }
+
+    @Override
+    protected void failedToConnect(Throwable exception) {
+        super.failedToConnect(exception);
+        closeConnection();
+    }
+
+    @Override
+    protected void connect() {
+        if (isConnected()) {
+            return;
+        }
+
+        if (client == null) {
+            client = clientBuilder.build();
+        }
+
+        client.start();
+
+        try {
+            LOG.info("Connecting to ZooKeeper with client: ", client);
+            client.blockUntilConnected();
+            LOG.info("Connected to ZooKeeper");
+        } catch (InterruptedException e) {
+            LOG.warn("Interrupted while connecting to ZooKeeper", e);
```

Comment:
Can we safely continue here or do we have to `return`?

Reply:
I don't think we can safely continue. I've modified this so that we will continue trying to blockUntilConnected() until it succeeds. This way if we get interrupted we will just go right back to trying.